prov/efa/src/efa_hmem.c

```diff
@@ -78,6 +80,106 @@ static int efa_domain_hmem_info_init_protocol_thresholds(struct efa_domain *efa_
 	return 0;
 }
 
+/**
+ * @brief Retrieve the P2P capability provider name for the chosen domain
+ * 
+ * @param[in]		efa_domain	EFA domain
+ * @param[in,out]	name		The output string
+ * @param[in]		len		The expected provider name length
+ * @return		On success, return the (positive) length of the
+ * 			provider name; otherwise a negative error code.
+ */
+static ssize_t efa_domain_hmem_p2p_prov_name(struct efa_domain *efa_domain,
+					     char *name, uint16_t len)
+{
+	FILE *fp = NULL;
+	char *ibdev_path, *p2p_path = NULL;
+	ssize_t ret = -FI_EINVAL;
+
+	ibdev_path = efa_domain->device->ibv_ctx->device->ibdev_path;
+	if (!ibdev_path) {
+		EFA_WARN(FI_LOG_DOMAIN, "IB device sysfs is not defined\n");
+		ret = -FI_EINVAL;
+		goto out;
+	}
+
+	p2p_path = malloc(strlen(ibdev_path) + strlen(efa_env.p2p_file_suffix) + 1);
+	if (!p2p_path) {
+		ret = -FI_ENOMEM;
+		goto out;
+	}
+
+	strcpy(p2p_path, ibdev_path);
+	strcat(p2p_path, efa_env.p2p_file_suffix);
+
+	fp = fopen(p2p_path, "r");
+	if (!fp) {
+		EFA_WARN(FI_LOG_DOMAIN, "Cannot open P2P file: %s\n", p2p_path);
+		ret = -FI_ENOENT;
+		goto out;
+	}
+
+	ret = (ssize_t) fread(name, 1, (size_t) len, fp);
+	if (ret <= 0) {
+		EFA_WARN(FI_LOG_DOMAIN, "P2P provider is not available\n");
+		ret = -FI_ENOSYS;
+		goto out;
+	} else if (ret >= (ssize_t) len) {
+		EFA_WARN(FI_LOG_DOMAIN, "Truncated P2P provider name\n");
+		ret = -FI_ETRUNC;
+		goto out;
+	}
+
+	name[ret] = '\0';
+	EFA_INFO(FI_LOG_DOMAIN, "P2P provider name: %s\n", name);
+out:
+	if (fp) {
+		fclose(fp);
+	}
+	if (p2p_path) {
+		free(p2p_path);
+	}
+	return ret;
+}
+
+/**
+ * @brief Determine if peer-to-peer capability is supported on the domain
+ *
+ * @param[in]	efa_domain	EFA domain
+ * @param[in]	iface		HMEM interface to check
+ * @return	true if P2P is supported, otherwise false
+ */
+static bool efa_domain_hmem_support_p2p(struct efa_domain *efa_domain, enum fi_hmem_iface iface)
+{
+	bool ret = true;
+	char p2p_prov[P2P_PROV_MAX_LEN];
+	const char *p2p_prov_prefix;
+
+	if (iface == FI_HMEM_SYSTEM || iface == FI_HMEM_SYNAPSEAI) {
+		goto out;
+	}
+
+	assert(iface == FI_HMEM_CUDA || iface == FI_HMEM_NEURON);
+
+	p2p_prov_prefix = iface == FI_HMEM_CUDA ? "NVIDIA" : "NEURON";
+
+	if ((efa_domain_hmem_p2p_prov_name(efa_domain, p2p_prov,
+					   P2P_PROV_MAX_LEN)) <= 0) {
+		EFA_INFO(FI_LOG_DOMAIN, "Failed to get P2P provider\n");
+		ret = false;
+		goto out;
+	}
+
+	if (strlen(p2p_prov) < strlen(p2p_prov_prefix) ||
+	    strncmp(p2p_prov, p2p_prov_prefix, strlen(p2p_prov_prefix))) {
+		EFA_INFO(FI_LOG_DOMAIN, "P2P provider does not support hmem interface: %d\n", iface);
+		ret = false;
+		goto out;
+	}
+out:
```

Comment:
do you really need an out here? shouldn't it just need `return ret`?

Reply:
Cool please see update.